### PR TITLE
Bulk session creation

### PR DESCRIPTION
**Pull request by sumanvpacewisdom**

### CODE REVIEW

File: src/database/migrations/20240404102919-upload_session_role_permissions.js [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+module.exports = {
+	async up(queryInterface, Sequelize) {
+		try {
+			const rolePermissionsData = [
+				{
+					role_title: common.MENTEE_ROLE,
+					permission_id: await getPermissionId(
+						'cloud-services',
+						['POST', 'DELETE', 'GET', 'PUT', 'PATCH'],
+						'/mentoring/v1/cloud-services/*'
```
Comment (rakeshSgr) [inline src/database/migrations/20240404102919-upload_session_role_permissions.js:33]: @sumanvpacewisdom Why this permission required?

Comment (sumanvpacewisdom) [reply src/database/migrations/20240404102919-upload_session_role_permissions.js:33]: only session manager can create the session bulk upload right , so i have given permission only for session manager role

File: src/scripts/sessionUploadScript.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
+const fs = require('fs')
+require('dotenv').config({ path: './.env' })
+const path = require('path')
+const fileService = require('@services/files')
+const request = require('request')
+const common = require('@constants/common')
+
+;(async () => {
+	try {
+		const fileName = 'sample.csv'
```
Comment (rakeshSgr) [inline src/scripts/sessionUploadScript.js:10]: @sumanvpacewisdom Rename to bulkSessionCreation.csv

File: src/services/org-admin.js [hunk 3]
```diff
@@ -580,4 +583,69 @@ module.exports = class OrgAdminService {
...
+	 * @name bulkUserCreate
+	 * @param {Array} users - user details.
+	 * @param {Object} tokenInformation - token details.
+	 * @returns {CSV} - created users.
+	 */
+
+	static async bulkSessionCreate(filePath, tokenInformation) {
+		try {
+			const { id, organization_id } = tokenInformation
+			const userDetail = await userRequests.details('', id)
```
Comment (rakeshSgr) [inline src/services/org-admin.js:599]: @sumanvpacewisdom Add this call after validating the csv input.

Comment (sumanvpacewisdom) [reply src/services/org-admin.js:599]: @rakeshSgr this is the details of the user who is uploading the csv not details of csv and also the csv validation is done after sending it to queue.

File: src/services/sessionInvite.js [hunk 4]
```diff
@@ -0,0 +1,422 @@
...
+			}
+		} catch (error) {
+			return {
+				success: false,
+				message: error.message,
+			}
+		}
+	}
+
+	static async sendInviteeEmail(templateData, userData, inviteeUploadURL = null, subjectComposeData = {}) {
```
Comment (rakeshSgr) [inline src/services/sessionInvite.js:391]: @sumanvpacewisdom Change the appropriate name

File: src/services/sessionInvite.js [hunk 5]
```diff
@@ -0,0 +1,422 @@
...
+					session.status = 'Invalid data: Compulsory field(s) missing'
+					rowsWithStatus.push(session)
+					invalidRowsCount++
+				} else {
+					session.status = 'Valid'
+					validRowsCount++
+					session.start_date = utils.convertToEpochTime(session.date, session.time24hrs, session.time_zone)
+					const durationTime = utils.addDurationToTime(session.time24hrs, session.duration)
+					session.end_date = utils.convertToEpochTime(session.date, durationTime, session.time_zone)
+					if (session.mentees.length === 0 || session.mentees === '[]') {
```
Comment (rakeshSgr) [inline src/services/sessionInvite.js:250]: @sumanvpacewisdom remove this

File: src/services/sessionInvite.js [hunk 6]
```diff
@@ -0,0 +1,422 @@
...
+					} else {
+						session.status = 'Invalid Mentor Email'
+						validRowsCount--
+						invalidRowsCount++
+					}
+
+					rowsWithStatus.push(session)
+				}
+			}
+			const outputArray = rowsWithStatus.map((item) => ({
```
Comment (rakeshSgr) [inline src/services/sessionInvite.js:283]: @sumanvpacewisdom use proper naming for variables

File: src/services/sessionInvite.js [hunk 7]
```diff
@@ -0,0 +1,422 @@
...
+			return {
+				success: false,
+				message: error,
+			}
+		}
+	}
+	static async processCsvData(dataArray, userId, orgId, isMentor, notifyUser) {
+		const output = []
+		for (const data of dataArray) {
+			const sessionCreation = await sessionService.create(data, userId, orgId, isMentor, notifyUser)
```
Comment (rakeshSgr) [inline src/services/sessionInvite.js:328]: @sumanvpacewisdom Can we make this

File: src/services/sessionInvite.js [hunk 8]
```diff
@@ -0,0 +1,422 @@
...
+					!session.description &&
+					!session.date &&
+					!session.time24hrs &&
+					!session.medium &&
+					!session.type &&
+					!session.mentor_id &&
+					!session.duration &&
+					!session.time_zone
+				) {
+					session.status = 'Invalid data: Compulsory field(s) missing'
```
Comment (rakeshSgr) [inline src/services/sessionInvite.js:241]: @sumanvpacewisdom check the story update the error message as per that

File: src/bulkSessionCreation.csv [hunk 9]
```diff
@@ -0,0 +1,5 @@
+Session Title,Description,Session Type,Mentor(Email/Mobile Num),Mentees(Email/Mobile Num),Date(DD-MM-YYYY),Time Zone(IST/UTC),Time (24 hrs),Duration(Min),Recommended For,Categories,Medium,Meeting Link,Meeting Id (if needed),Meeting Passcode (if needed)
+School Improvement,Session on school improvement ideas,Private,menteeoneprod@grr.la,"suman.v@gmail.com, user1@sharklasers.com",20-04-2024,IST,13:00 Hrs,60,"Headmasters","Infrastucture","English",
+Session Math,Session on Pedogogy,Public,gagan123@gmail.com,"",20-04-2024,UTC,09:00 Hrs,90,"Teachers","Pedagogy","Hindi",https://meet.google.com/jgx-ydjs-ptv?authuser=0
+Communication Skills,Basics of Communcation Skills,Public,shiva@shark.com,"alok@gmail.com, Amey1234@org1.com, afnan.lj@org2.in",20-04-2024,IST,11:30 Hrs,30,"","","English",https://zoom.us/j/5551112222,5551112222,12345
+Mahakavi Kuvempu,Life History of Mahakavi Kuvempu,Public,shiva@shark.com,"",20-04-2024,IST,11:30 Hrs,120,"Teachers","","Kannada",https://call.whatsapp.com/video/tK0CUX6h4AjDsWoQNStz1g
```
Comment (nevil-mathew) [inline src/bulkSessionCreation.csv:5]: @sumanvpacewisdom can we have CSVs under /assets folder ?

File: src/controllers/v1/org-admin.js [hunk 10]
```diff
@@ -144,4 +145,36 @@ module.exports = class OrgAdmin {
...
+	}
+
+	/**
+	 * Upload sample csv based on org id
+	 * @method
+	 * @name UploadSampleCsv
+	 * @param {String} req.body.file_path -Uploaded filr path .
+	 * @returns {Object} - uploaded file response.
+	 */
+	async bulkSessionCreate(req) {
```
Comment (nevil-mathew) [inline src/controllers/v1/org-admin.js:172]: Why does this endpoint have the same name as above?

File: src/services/files.js [hunk 11]
```diff
@@ -74,4 +76,82 @@ module.exports = class FilesHelper {
 			throw error
 		}
 	}
+
+	static async getSignedUrlByOrgId(fileName, id, orgId, dynamicPath) {
```
Comment (nevil-mathew) [inline src/services/files.js:80]: Why do we need this function?

File: src/services/files.js [hunk 12]
```diff
@@ -74,4 +76,82 @@ module.exports = class FilesHelper {
...
+				statusCode: httpStatusCode.ok,
+				responseCode: 'OK',
+				result: response,
+			})
+		} catch (error) {
+			throw error
+		}
+	}
+
+	static async getDownloadableUrlByOrgId(orgId) {
```
Comment (nevil-mathew) [inline src/services/files.js:124]: Why do we need this function, why aren't we reusing the existing downloadUrl()?

File: src/services/sessions.js [hunk 13]
```diff
@@ -106,7 +106,7 @@ module.exports = class SessionsHelper {
 			// If time slot not available return corresponding error
 			if (timeSlot.isTimeSlotAvailable === false) {
 				const errorMessage = isSessionCreatedByManager
-					? 'INVALID_TIME_SELECTION_FOR_GIVEN_MENTOR'
+					? 'SESSION_CREATION_LIMIT_EXCEDED_FOR_GIVEN_MENTOR'
```
Comment (nevil-mathew) [inline src/services/sessions.js:109]: Why was this message changed ?

File: src/services/sessions.js [hunk 14]
```diff
@@ -151,6 +151,7 @@ module.exports = class SessionsHelper {
 					statusCode: httpStatusCode.bad_request,
 					responseCode: 'CLIENT_ERROR',
 				})
+			console.log('================================', defaultOrgId)
```
Comment (nevil-mathew) [inline src/services/sessions.js:154]: remove this console.log()

File: src/services/sessions.js [hunk 15]
```diff
@@ -1755,7 +1770,7 @@ module.exports = class SessionsHelper {
 	static async isTimeSlotAvailable(id, startDate, endDate, sessionId) {
 		try {
 			const sessions = await sessionQueries.getSessionByUserIdAndTime(id, startDate, endDate, sessionId)
-			if (!sessions) {
+			if (!sessions || sessions.startDateResponse.length < process.csv.SESSION_CREATION_MENTOR_LIMIT) {
```
Comment (nevil-mathew) [inline src/services/sessions.js:1773]: Where are you getting this '**startDateResponse**' from ?
Also why is it handled in this function which is specific to isTimeSlotAvailable()

File: src/package.json [hunk 16]
```diff
@@ -32,6 +32,7 @@
 		"crypto": "^1.0.1",
 		"deep-object-diff": "^1.1.9",
 		"dotenv": "^10.0.0",
+		"elevate-cloud-storage": "^2.6.3",
```
Comment (nevil-mathew) [inline src/package.json:35]: This was removed right?

File: src/constants/common.js [hunk 17]
```diff
@@ -123,6 +123,21 @@ module.exports = {
...
+	sessionOutputFile: 'output-session-creation',
+	csvExtension: '.csv',
+	responseType: 'stream',
+	azureBlobType: 'BlockBlob',
+	UPLOADED_STATUS: 'UPLOADED',
+	FAILED_STATUS: 'FAILED',
+	PROCESSED_STATUS: 'PROCESSED',
+	notificationEmailType: 'email',
+	STATUSCODE: 201,
+	CREATEDRESPONSECODE: 'OK',
```
Comment (nevil-mathew) [inline src/constants/common.js:139]: Why do we have these response codes etc in common.js? Instead of generics/http-status.js

File: src/controllers/v1/cloud-services.js [hunk 18]
```diff
@@ -16,9 +16,10 @@ module.exports = class CloudServices {
 	 */
 	async getSignedUrl(req) {
 		try {
-			const signedUrlResponse = await filesService.getSignedUrl(
+			const signedUrlResponse = await filesService.getSignedUrlByOrgId(
```
Comment (nevil-mathew) [inline src/controllers/v1/cloud-services.js:19]: Why was this changed from getSignedUrl to getSignedUrlByOrgId()?

File: src/constants/common.js [hunk 19]
```diff
@@ -123,6 +123,21 @@ module.exports = {
...
+	NO_OF_ATTEMPTS: 3,
+	BACK_OFF_RETRY_QUEUE: 600000,
+	tempFolderForBulkUpload: 'public/invites',
+	sessionOutputFile: 'output-session-creation',
+	csvExtension: '.csv',
+	responseType: 'stream',
+	azureBlobType: 'BlockBlob',
+	UPLOADED_STATUS: 'UPLOADED',
+	FAILED_STATUS: 'FAILED',
+	PROCESSED_STATUS: 'PROCESSED',
```
Comment (nevil-mathew) [inline src/constants/common.js:136]: can we have in 
`{
  "STATUS": {
    "FAILED": "FAILED",
    "PROCESSED": "PROCESSED",
    "UPLOADED": "UPLOADED"
  }
}
`
this format.

File: src/controllers/v1/org-admin.js (hunk 10 above)
Comment (sumanvpacewisdom) [reply src/controllers/v1/org-admin.js:172]: @nevil-mathew I will remove it , i have missed it out . We need only one here

File: src/services/files.js (hunk 11 above)
Comment (sumanvpacewisdom) [reply src/services/files.js:80]: @nevil-mathew here we are getsignedUrl based on orgId , if they need to create a custom csv for particular
orgId , i am using this function.

File: src/services/files.js (hunk 12 above)
Comment (sumanvpacewisdom) [reply src/services/files.js:124]: @nevil-mathew This is needed to download csv based on orgId , if there is csv present for that orgId it will download the custom csv else the default csv

File: src/services/sessions.js (hunk 13 above)
Comment (sumanvpacewisdom) [reply src/services/sessions.js:109]: @nevil-mathew So we are introducing the new flow where 5 sessions can be created for one mentor for same time . where he can accept one and reject rest .

File: src/services/sessions.js (hunk 15 above)
Comment (sumanvpacewisdom) [reply src/services/sessions.js:1773]: @nevil-mathew We are querying it from sessions table and we have a custom response for the start and end Date
So we are introducing the new flow where 5 sessions can be created for one mentor for same time . where he can accept one and reject rest .

File: src/package.json (hunk 16 above)
Comment (sumanvpacewisdom) [reply src/package.json:35]: @nevil-mathew I am not sure , do we have to remove it , i merged the 3.0 release and in this file ,there was a conflict here , i accepted both changes .

File: src/constants/common.js (hunk 17 above)
Comment (sumanvpacewisdom) [reply src/constants/common.js:139]: @nevil-mathew will use it from generics file .

File: src/controllers/v1/cloud-services.js (hunk 18 above)
Comment (sumanvpacewisdom) [reply src/controllers/v1/cloud-services.js:19]: @nevil-mathew Intially we were upload genric file for all the organizations , but the flow is changed to each organization can have it own custom csv.

File: src/constants/common.js (hunk 19 above)
Comment (sumanvpacewisdom) [reply src/constants/common.js:136]: @nevil-mathew Will have to look into it , how it works .

File: src/generics/utils.js [hunk 20]
```diff
@@ -552,6 +591,122 @@ function validateFilters(input, validationData, modelName) {
 	return input
 }
 
+const algorithm = process.env.EMAIL_ID_ENCRYPTION_ALGORITHM
+const secretKey = Buffer.from(process.env.EMAIL_ID_ENCRYPTION_KEY, 'hex')
+const fixedIV = Buffer.from(process.env.EMAIL_ID_ENCRYPTION_IV, 'hex')
+
+const encrypt = (plainTextEmail) => {
```
Comment (nevil-mathew) [inline src/generics/utils.js:598]: These cannot be in mentoring service. The encryption and decryption should be handled by user service.

File: src/generics/utils.js [hunk 21]
```diff
@@ -552,6 +591,122 @@ function validateFilters(input, validationData, modelName) {
...
+	let minutes = parseInt(minutesStr, 10)
+	// Add the duration to the hours and minutes
+	hours += Math.floor(durationMinutes / 60)
+	minutes += durationMinutes % 60 // Remainder gives updated minutes
+	// Format the updated time
+	const updatedTimeString = `${hours.toString().padStart(2, '0')}:${minutes.toString().padStart(2, '0')}`
+	return updatedTimeString
+}
+
+function convertToEpochTime(dateString, timeString, timeZone) {
```
Comment (nevil-mathew) [inline src/generics/utils.js:679]: Any reason to have a completly new function for this?
Cant this be handled by moment itself?

File: src/generics/utils.js (hunk 20 above)
Comment (sumanvpacewisdom) [reply src/generics/utils.js:598]: @nevil-mathew Will change it .

File: src/generics/utils.js (hunk 21 above)
Comment (sumanvpacewisdom) [reply src/generics/utils.js:679]: @nevil-mathew so we have Time timeZone and and duration separatly in csv , in portal front end will convert the time to epoach and will send it to backend , so here its only backend and we need to convert to epoach time , so we are using this .

File: src/services/files.js (hunk 11 above)
Comment (nevilmathew) [reply src/services/files.js:80]: What does signedURL have to do with org id?

File: src/generics/utils.js (hunk 21 above)
Comment (nevilmathew) [reply src/generics/utils.js:655]: Sure i got that, but what i was asking is the need for separate function?
Conversion can be easily handled moment().unix() right?

File: src/generics/utils.js (hunk 20 above)
Comment (nevilmathew) [reply src/generics/utils.js]: User service always accepts plain text email right?
You can ref /login

File: src/generics/utils.js (hunk 21 above)
Comment (nevil-mathew) [reply src/generics/utils.js:655]: Sure got that, but what I was asking is the need for separate function?
Conversion can be easily handled moment().unix() right?

File: src/generics/utils.js (hunk 20 above)
Comment (nevil-mathew) [reply src/generics/utils.js]: User service always accepts plain text email, right?
You can ref /login

File: src/generics/utils.js (hunk 21 above)
Comment (sumanvpacewisdom) [reply src/generics/utils.js:655]: @nevil-mathew will have to look into it , and will change if it is possible

File: src/generics/utils.js (hunk 20 above)
Comment (sumanvpacewisdom) [reply src/generics/utils.js]: @nevil-mathew I have chnaged it and , i am sending plainTextEmail to user service and encrypting in user services.

File: src/database/migrations/20240430072700-update_csv_filepath_column_org_extention.js [hunk 22]
```diff
@@ -0,0 +1,13 @@
+'use strict'
+/** @type {import('sequelize-cli').Migration} */
+module.exports = {
+	async up(queryInterface, Sequelize) {
+		await queryInterface.addColumn('organization_extension', 'sample_csv_path', {
+			type: Sequelize.STRING,
```
Comment (nevil-mathew) [inline src/database/migrations/20240430072700-update_csv_filepath_column_org_extention.js:6]: Use JSONB, rename to uploads

Comment (nevil-mathew) [reply src/database/migrations/20240430072700-update_csv_filepath_column_org_extention.js:6]: {
    session_sample_csv_path
}
